Services/src/main/java/ev/projects/services/LogService.java

```diff
@@ -0,0 +1,58 @@
+package ev.projects.services;
+
+import ev.projects.models.Log;
+import ev.projects.repositories.ILogRepository;
+import org.apache.commons.csv.CSVFormat;
+import org.apache.commons.csv.CSVPrinter;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Service;
+import java.io.IOException;
+import java.io.PrintWriter;
+import java.util.List;
+import java.util.Optional;
+
+@Service
+public class LogService implements ILogService {
+
+    private ILogRepository logRepository;
+
+    @Autowired
+    public LogService(ILogRepository logRepository) {
+        this.logRepository = logRepository;
+    }
+
+    @Override
+    public List<Log> getAll() {
+        return logRepository.findAll();
+    }
+
+    @Override
+    public Optional<Log> getById(long ID) {
+        return logRepository.findById(ID);
+    }
+
+    @Override
+    public void add(Log log) {
+        logRepository.save(log);
+    }
+
+    @Override
+    public void getLog(long ID, PrintWriter writer) {
+        logRepository.findById(ID).map(l -> {
+            extractDataAsCsv(List.of(l), writer);
+            return null;
+        });
+    }
+
+    private void extractDataAsCsv(List<Log> logs, PrintWriter writer) {
+        try (CSVPrinter csvPrinter = new CSVPrinter(writer, CSVFormat.DEFAULT)) {
```

Comment:
> Is it necessary to create new instance of csvPrinter in every iteration of a loop if it uses same writer? Maybe worth creating it in getLog() method and pass it to extractDataAsCsv()

If I see this correctly, csvPrinter is created outside of the for loop so it is created only once, I think.
The method who calls the exctractDataAsCsv method uses findById, therefore -> only one call. If findAll was used it would be passed directly as argument to the method.
Am I missing something?

Reply:
> also, good practice to call close() on a printer in finally block

I can call close, but as I have read, since printer is initialized in try block header, it is closed automatically, as mentioned in: https://stackoverflow.com/questions/29650468/is-closing-resources-in-try-block-accepted. Does this approach has any flaws?